### PR TITLE
[SYCL] Report an error message when "sycl::program::create_program_with_source" is used with Level-Zero

### PR DESCRIPTION
**Pull request by rbegam**
Level-Zero backend does not support create_program_with_source. This is to report an error message from SYCL RT if that is requested.

Signed-off-by: rbegam <rehana.begam@intel.com>

### CODE REVIEW

File: sycl/source/detail/program_impl.cpp [hunk 1]
```diff
@@ -358,8 +358,16 @@ void program_impl::create_cl_program_with_source(const string_class &Source) {
...
-  Plugin.call<PiApiKind::piclProgramCreateWithSource>(
-      MContext->getHandleRef(), 1, &Src, &Size, &MProgram);
+  RT::PiResult Err =
+      Plugin.call_nocheck<PiApiKind::piclProgramCreateWithSource>(
+          MContext->getHandleRef(), 1, &Src, &Size, &MProgram);
+
+  if (Plugin.getBackend() == cl::sycl::backend::level_zero &&
+      Err == PI_INVALID_OPERATION) {
+    throw feature_not_supported(
+        "piclProgramCreateWithSource is not supported in Level Zero",
```
Comment (smaslov-intel) [inline sycl/source/detail/program_impl.cpp:368]:
```suggestion
        "program::compile_with_source is not supported with Level Zero backend.",
```

We must not tell end users about PI interfaces. Also I think it would be good to redirect to the library that can be used to build OpenCL kernel for Level-Zero. @v-klochkov , which one is this?

**Review by smaslov-intel — CHANGES_REQUESTED** [review]

File: sycl/source/detail/program_impl.cpp (hunk 1 above)
Comment (bader) [reply sycl/source/detail/program_impl.cpp:368]: Shouldn't this error be emitted from Level Zero plug-in, rather than from DPC++ runtime?

Comment (smaslov-intel) [reply sycl/source/detail/program_impl.cpp:368]: Plugins are C API. In order to throw a consistent error message this is better suited for SYCL RT, I think.

Comment (bader) [reply sycl/source/detail/program_impl.cpp:368]: > In order to throw a consistent error message this is better suited for SYCL RT, I think.

Right, but BE specific error detection logic must be located in corresponding plug-in. DPC++ runtime is designed to keep device specific logic in plug-ins and logic common for all back-ends in "runtime" library.
To follow this design we can separate error detection logic (device specific) from error reporting/conversion logic (common for all plug-ins).
E.g. plug-in can return an error code when "compile with source" feature is not supported (PI_INVALID_OPERATION? or some other error code) and DPC++ runtime can report this error via exception - "program::compile_with_source is not supported with" + Plugin.getBackend().to_string()" - this is a pseudo code to demonstrate the idea, not a real C++ program.

Basically code like this: 
```C++
if (Plugin.getBackend() == cl::sycl::backend::<some_backend>) {
...
}
```
is a red flag signaling that most likely it should be implemented in corresponding plug-in.

Comment (smaslov-intel) [reply sycl/source/detail/program_impl.cpp:368]: That's right. We don't need to check the backend here in SYCL RT, every backend would return PI_INVALID_OPERATION of compilation is not supported, and the error message would be common for all such backends. @rbegam , please also change CUDA plugin to do the same, i.e. return an error: https://github.com/intel/llvm/blob/eab4791cf727ac889857d4723c5c9284489c9537/sycl/plugins/cuda/pi_cuda.cpp#L2653

Comment (v-klochkov) [reply sycl/source/detail/program_impl.cpp:368]: > Also I think it would be good to redirect to the library that can be used to build OpenCL kernel for Level-Zero. @v-klochkov , which one is this?

The class sycl::INTEL::online_compiler can generate IL/SPIR-V for a source program.

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 2]
```diff
@@ -2650,7 +2650,7 @@ pi_result cuda_piclProgramCreateWithSource(pi_context context, pi_uint32 count,
                                            const size_t *lengths,
                                            pi_program *program) {
   cl::sycl::detail::pi::die("cuda_piclProgramCreateWithSource not implemented");
```
Comment (smaslov-intel) [inline sycl/plugins/cuda/pi_cuda.cpp:2652]: remove the die, since it makes the return unreachable

File: sycl/source/detail/program_impl.cpp [hunk 3]
```diff
@@ -26,6 +27,21 @@ __SYCL_INLINE_NAMESPACE(cl) {
 namespace sycl {
 namespace detail {
 
+inline const std::string GetBackendString(cl::sycl::backend backend) {
```
Comment (smaslov-intel) [inline sycl/source/detail/program_impl.cpp:30]: if you really wanted to name the backend in the error message (I don't feel this is necessary) then you'd employ the existing https://github.com/intel/llvm/blob/cc3d0956c46135b2791d9adaf191fe82ff67eb8e/sycl/include/CL/sycl/backend_types.hpp#L31

**Review by smaslov-intel — CHANGES_REQUESTED** [review]

File: sycl/plugins/cuda/pi_cuda.cpp (hunk 2 above)
Comment (rbegam) [reply sycl/plugins/cuda/pi_cuda.cpp:2652]: right, will do that. should I enable printing the error msg here (like level_zero) before returning? There are some more cases in the cuda plugin where die() is called before a return stmt with an error code. might be a good idea to fix those too.

Comment (smaslov-intel) [reply sycl/plugins/cuda/pi_cuda.cpp:2652]: >  should I enable printing the error msg here (like level_zero) before returning?

Why not, if they have the ways to print logs.

> There are some more cases in the cuda plugin where die() is called before a return stmt with an error code. might be a good idea to fix those too.

Let's not remove any "die"-s without making sure they (error return codes) are handled properly in SYCL RT.

Comment (rbegam) [reply sycl/plugins/cuda/pi_cuda.cpp:2652]: > Why not, if they have the ways to print 

there is no way right now just for printing error msg.

File: sycl/source/detail/program_impl.cpp (hunk 3 above)
Comment (rbegam) [reply sycl/source/detail/program_impl.cpp:30]: I'll skip this if not necessary.

**Review by vladimirlaz — CHANGES_REQUESTED** [review]
@rbegam, could you please fix the clang-format issue

File: sycl/source/detail/program_impl.cpp [hunk 4]
```diff
@@ -358,8 +358,14 @@ void program_impl::create_cl_program_with_source(const string_class &Source) {
...
   size_t Size = Source.size();
   const detail::plugin &Plugin = getPlugin();
-  Plugin.call<PiApiKind::piclProgramCreateWithSource>(
-      MContext->getHandleRef(), 1, &Src, &Size, &MProgram);
+  RT::PiResult Err =
+      Plugin.call_nocheck<PiApiKind::piclProgramCreateWithSource>(
+          MContext->getHandleRef(), 1, &Src, &Size, &MProgram);
+
+  if (Err == PI_INVALID_OPERATION) {
+    throw feature_not_supported("program::compile_with_source is not supported",
```
Comment (smaslov-intel) [inline sycl/source/detail/program_impl.cpp:366]:
```suggestion
    throw feature_not_supported("program::compile_with_source is not supported by the selected backend",
```

**Review by smaslov-intel — DISMISSED** [review]

**Review by smaslov-intel — APPROVED** [review]
LGTM

**Review by smaslov-intel — APPROVED** [review]

**Review by vladimirlaz — APPROVED** [review]
